File: lectures/L09/flatten_layer/cpp/flatten_demo.cpp
/**
 * @file Simple flatten layer demo.
 */
#include <cstdio>
#include <stdexcept>
#include <vector>

namespace
{
/** One-dimensional matrix. */
using Matrix1d = std::vector<double>;

/** Two-dimensional matrix. */
using Matrix2d = std::vector<Matrix1d>;

/**
 * @brief Initialize matrix with zeros.
 *
 * @param[in] matrix The matrix to initialize.
 */
void initMatrix(Matrix1d& matrix) noexcept
{
    // Fill the matrix with zeros.
    for (auto& num : matrix)
    {
        num = 0.0;
    }
}

/**
 * @brief Initialize matrix with zeros.
 *
 * @param[in] matrix The matrix to initialize.
 */
void initMatrix(Matrix2d& matrix) noexcept
{
    // Fill the matrix with zeros.
    for (auto& row : matrix)
    {
        for (auto& num : row)
        {
            num = 0.0;
        }
    }
}

/**
 * @brief Initialize matrix with zeros.
 *
 * @param[in] matrix The matrix to initialize.
 * @param[in] size The desired size of the matrix.
 */
void initMatrix(Matrix1d& matrix, const std::size_t size)
{
    // Resize the matrix if necessary, then fill the matrix with zeros.
    matrix.resize(size);
    initMatrix(matrix);
}

/**
 * @brief Initialize matrix with zeros.
 *
 * @param[in] matrix The matrix to initialize.
 * @param[in] size The desired size of the matrix.
 */
void initMatrix(Matrix2d& matrix, const std::size_t size)
{
    // Resize the matrix if necessary, then fill the matrix with zeros.
    matrix.resize(size, Matrix1d(size));
    initMatrix(matrix);
}

/**
 * @brief Check whether given matrix is square.
 *
 * @param[in] matrix The matrix to check.
 *
 * @return True if the matrix is square, false otherwise.
 */
[[nodiscard]] bool isMatrixSquare(const Matrix2d& matrix) noexcept
{
    // Check the matrix row by row.
    for (const auto& row : matrix)
    {
        // Return false if the matrix isn't square.
        if (row.size() != matrix.size()) { return false; }
    }
    // Return true to indicate that the matrix is square.
    return true;
}

/**
 * @brief Print the contents of given matrix.
 *
 * @param[in] matrix The matrix to print.
 */
void printMatrix(const Matrix1d& matrix) noexcept
{
    const auto* lastNum{matrix.empty() ? nullptr : &matrix[matrix.size() - 1U]};

    // Print the matrix contents on a single row.
    for (const auto& num : matrix)
    {
        // Separate each number in the row with a comma.
        std::printf("%.1f", num);
        if (&num < lastNum) { std::printf(", "); }
    }
    std::printf("\n");
}

/**
 * @brief Print the contents of given matrix.
 *
 * @param[in] matrix The matrix to print.
 */
void printMatrix(const Matrix2d& matrix) noexcept
{
    // Print the matrix contents row by row.
    for (const auto& row : matrix)
    {
        const auto* lastNum{row.empty() ? nullptr : &row[row.size() - 1U]};
        std::printf("\t");

        // Separate each number in the row with a comma.
        for (const auto& num : row)
        {
            std::printf("%.1f", num);
            if (&num < lastNum) { std::printf(", "); }
        }
        std::printf("\n");
    }
    std::printf("\n");
}

namespace ml
{
/**
 * @brief Flatten layer structure.
 *
 * @todo Implement this struct!
 */
struct FlattenLayer final
{
    /**
     * @brief Constructor.
     *
     * @param[in] inputSize Input size. Must be greater than 0.
     */

    /**
     * @todo Check the input size in the constructor:
     *
     *       If the input size is 0, terminate the program with error message
     *       "Cannot create flatten layer: invalid input size!".
     */

    /**
     * @brief Flatten the input from 2D to 1D.
     *
     * @param[in] input Matrix holding input data.
     *
     * @return True on success, false on failure.
     */

    /**
     * @brief Unflatten the output gradients from 1D to 2D.
     *
     * @param[in] outputGradients Matrix holding output gradients.
     *
     * @return True on success, false on failure.
     */

    /** @todo Define member variables here! Keep them public and skip the 'my' prefix. */
};
} // namespace ml
} // namespace

/**
 * @brief Create and demonstrate a simple flatten layer.
 *
 * @return 0 on success, -1 on failure.
 */
int main()
{
//! @todo Remove this header guard (and/or uncomment the compiler flags in the makefile) once the
//        implementation is finished.
#ifdef FLATTEN_LAYER_IMPLEMENTED
    // Example 4x4 input matrix (could represent an image or feature map).
    const Matrix2d input{{2, 1, 6, 1}, {3, 0, 4, 6}, {1, 2, 4, 5}, {3, 4, 7, 7}};

    // Example output gradients (same shape as flattened output, used for backpropagation demo).
    const Matrix1d outputGradients{1, 2, 3, 4, 8, 7, 6, 5, 0, 2, 4, 8, 9, 7, 5, 3};

    // Create a flatten layer: 4x4 input, produces 1x16 output.
    constexpr std::size_t inputSize{4U};
    ml::FlattenLayer flattenLayer{inputSize};

    // Perform feedforward (flatten the input), print the result.
    std::printf("Flattening input data (2D -> 1D):\n");
    printMatrix(input);
    flattenLayer.feedforward(input);
    std::printf("Resulting flattened output (1D):\n");
    printMatrix(flattenLayer.output);

    // Perform backpropagation (unflatten the output), print the result.
    std::printf("\nApplying backpropagation (1D -> 2D):\n");
    printMatrix(outputGradients);
    flattenLayer.backpropagate(outputGradients);
    std::printf("\nResulting unflattened input gradients (2D):\n");
    printMatrix(flattenLayer.inputGradients);
    return 0;

//! @todo Remove this header guard (and/or uncomment the compiler flags in the makefile) once the
//        implementation is finished.
#endif /** FLATTEN_LAYER_IMPLEMENTED */
}

File: lectures/L09/flatten_layer/python/flatten_demo.py
"""Simple flatten layer demo."""

# pylint: disable = consider-using-enumerate

# One-dimensional matrix.
Matrix1d = list[float]

# Two-dimensional matrix.
Matrix2d = list[Matrix1d]


def create_matrix1d(size: int) -> Matrix1d:
    """Create a one-dimensional matrix.

    Args:
        size: Size of the matrix.

    Returns:
        New matrix initialized with zeros.
    """
    return [0.0] * size


def create_matrix2d(size: int) -> Matrix2d:
    """Create a two-dimensional matrix.

    Args:
        size: Size of the matrix.

    Returns:
        New matrix initialized with zeros.
    """
    matrix: Matrix2d = []
    for _ in range(size):
        row = [0.0] * size
        matrix.append(row)
    return matrix


def is_matrix2d_square(matrix: Matrix2d) -> bool:
    """Check whether matrix is square.

    Args:
        matrix: The matrix to check.

    Returns:
        True if the matrix is square, false otherwise.
    """
    return all(len(row) == len(matrix) for row in matrix)


class FlattenLayer:
    """Flatten layer implementation."""

    def __init__(self, input_size: int) -> None:
        # Check the input size, throw an exception if invalid.
        if input_size == 0:
            raise ValueError("Cannot create flatten layer: invalid input size!")

        # Initialize the matrices.
        self.input_gradients = create_matrix2d(input_size)
        self.output = create_matrix1d(input_size * input_size)

    def feedforward(self, input_data: Matrix2d) -> bool:
        """Perform feedforward operation.

        Args:
            input_data: Input data to flatten (2D -> 1D).

        Returns:
            True on success, false on failure.
        """
        # Compute the input size of the flatten layer.
        input_size = len(self.input_gradients)
        # Check the input dimensions, return false on failure.
        if len(input_data) != input_size or not is_matrix2d_square(input_data):
            return False

        # Flatten the input data.
        for i in range(input_size):
            for j in range(input_size):
                index = i * input_size + j
                self.output[index] = input_data[i][j]
        # Return true to indicate success.
        return True

    def backpropagate(self, output_gradients: Matrix1d) -> bool:
        """Perform backpropagation.

        Args:
            output_gradients: Output gradients to unflatten (1D -> 2D).

        Returns:
            True on success, false on failure.
        """
        input_size = len(self.input_gradients)
        output_size = len(self.output)

        # Check the output dimensions, return false on failure.
        if len(output_gradients) != output_size:
            return False

        # Unflatten the output gradients.
        for i in range(input_size):
            for j in range(input_size):
                index = i * input_size + j
                self.input_gradients[i][j] = output_gradients[index]
        # Return true to indicate success.
        return True


def main() -> None:
    """Create and demonstrate a simple flatten layer."""

    # Example 4x4 input matrix (could represent an image or feature map).
    input_data = [
        [2, 1, 6, 1],
        [3, 0, 4, 6],
        [1, 2, 4, 5],
        [3, 4, 7, 7],
    ]

    # Example output gradients (same shape as flattened output, used for backpropagation demo).
    output_gradients = [1, 2, 3, 4, 8, 7, 6, 5, 0, 2, 4, 8, 9, 7, 5, 3]

    # Create a flatten layer: 4x4 input, produces 1x16 output.
    flatten_layer = FlattenLayer(4)

    # Perform feedforward (flatten the input), print the result.
    print("Flattening input_data (2D -> 1D):")
    for row in input_data:
        print("  ", " ".join(f"{num:.1f}" for num in row))

    flatten_layer.feedforward(input_data)
    print("\nResulting flattened output (1D):")
    print("  ", " ".join(f"{num:.1f}" for num in flatten_layer.output))

    # Perform backpropagation (unflatten the output gradients), print the result.
    print("\nApplying backpropagation (1D -> 2D):")
    print("  Output gradients:", " ".join(f"{num:.1f}" for num in output_gradients))

    flatten_layer.backpropagate(output_gradients)
    print("\nResulting unflattened input gradients (2D):")
    for row in flatten_layer.input_gradients:
        print("  ", " ".join(f"{num:.1f}" for num in row))


# Invoke the main function if this is the startup script.
if __name__ == "__main__":
    main()
